doc/config.md
# GIN CLI configuration

The GIN client allows for certain aspects of its operation to be configured.
This is accomplished by specifying key-value pairs, in YAML format, in a file called `config.yml`.
The location of this file differs per platform ([see below](#config-file-location)).

In addition to this global configuration, the [git-annex filtering criteria](filtering.md) can be configured for individual repositories, by placing a file called `config.yml` at the root of the repository.

## Defaults

The following shows the default values for every configurable option
```yaml
bin:
  git: git
  gitannex: git-annex
  ssh: ssh

servers:
  gin:
    web:
      protocol: https
      host: web.gin.g-node.org
      port: 443

    git:
      host: gin.g-node.org
      port: 22
      user: git
      hostkey: "gin.g-node.org,141.84.41.216 ecdsa-sha2-nistp256 AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBE5IBgKP3nUryEFaACwY4N3jlqDx8Qw1xAxU2Xpt5V0p9RNefNnedVmnIBV6lA3n+9kT1OSbyqA/+SgsQ57nHo0="

annex:
    minsize: 10M
    exclude: []
```

### Description of the configuration values:
- bin: The bin section is used to specify the names and locations of external programs used by the GIN client. Generally, these are simply specified by the name of the executable, since it is assumed their locations can be found in the user's PATH. If this is not the case, the locations can be specified for each individual program.
    - git: The path to the git executable.
    - gitannex: The path to the git-annex executable.
    - ssh: The path to the ssh executable.
- servers: The servers section is used to define GIN servers that the client can interact with. By default, there is only one server configured called 'gin'.
  - gin: The default GIN server. By default it points to the official G-Node GIN server. This can be changed to work with locally deployed servers. Additional servers can be added with different names (called aliases).
      - protocol: The protocol (scheme) used by the server, typically `http` or `https`.
      - host: The web address of the server.
      - port: The port used by the server, typically `80` for `HTTP` or `443` for `HTTPS`.
  - git: The git section is used to specify the git address, port, username, and host key that the GIN web server is configured to use.
      - address: The git/ssh server address. This is often the same as the web (gin) address, but may be different, or have a different subdomain.
      - port: The ssh server port (typically `22`).
      - user: For most git servers this is simply the user `git`. This is the name of the server-side user that handles all remote git operations.
      - hostkey: The SSH key of the git server. The GIN client uses strict host key checking, so if this is not specified, or is specified incorrectly, git operations will not work. This key is different for each server installation.
- annex: The annex section is used to specify the [git-annex filtering criteria](filtering.md). This is the only configuration section that is read for **local** (per repository) configurations.
    - minsize: The minimum size of a file that should be added to the annex. All files smaller than this size are added to git instead.
    - exclude: Patterns or filenames that should be excluded from the annex. For example, the pattern `*.py` will exclude all Python source code files from the annex, adding them to git instead. Files which match a pattern are always excluded from the annex, even if they are above the minsize. Patterns should be specified as a list of strings, e.g., `["*.py", "*.md", "*.m"]`.


## Config file location

The location of the user global configuration file differs per platform:

- Windows: `%APPDATA%\g-node\gin\config.yml`
- macOS: `${HOME}/Library/Application Support/g-node/gin/config.yml`
- Linux: `${HOME}/.config/g-node/gin/config.yml`


CHANGELOG.md
# Changelog

**Beta** releases are not listed. Changes for beta releases are included in the next full release. Current changes are listed in the top **Unreleased** section.

## Unreleased

### Changes
- New commands
    - Local workflows:
        - `gin init`: Initialises a directory as a gin repository without creating a repository on a server.
        - `gin commit`: Records changes in the local repository without performing an upload and without requiring a remote or an internet connection.
    - Versioning
        - `gin version`: Rolls back files or directories to older versions. With the `--copy-to` flag, retrieves older files without overwriting the current version and copies them to a specific location.
    - Remotes
        - `gin add-remote`: Adds a remote to the current repository for uploading and downloading. This also brings support for using directory paths on the local filesystem as "remotes" (e.g., external hard drives, network attached storage), without the need to have a GIN server.
        - `gin remove-remote`: Removes a remote from the current repository.
        - `gin remotes`: Lists remotes configured for a repository and shows the default remote used for `gin upload`.
        - `gin use-remote`: Switches the repository's default upload remote.
        - The `gin upload` command now accepts a `--to` argument for uploading annexed content to a specific, non-default remote.
    - Servers
        - `gin add-server`: Adds a new GIN server to the global client configuration.
        - `gin remove-server`: Removes a GIN server from the global client configuration.
        - `gin servers`: Lists the configured servers and shows the default server used for web queries.
        - `gin use-server`: Switches the default server.
        - The `login`, `logout`, `create`, `info`, `keys`, `repos`, `repoinfo`, and `get` commands now accept a `--server` argument for querying or operating on a specific, non-default server.
- Progress bars for file operations: Some operations don't have a per-file progress (add, lock, unlock). There is no partial unlock state for a file, for instance. For these commands, the output shows the overall progress along with the number of total files that are being affected.
- Smaller logfiles: The log file is now limited to 1 MB. No more ever-growing logs.

### Bug fixes
- Fixed a bug where file tracking would register a type change in git when working with direct mode repositories (e.g., on Windows).
- Fixed a bug where the file status (from `gin ls`) was being incorrectly reported when working in direct mode (e.g., on Windows). Direct mode repositories should now show the exact same output as indirect mode ones.

### Relevant PRs
- #191: New command: gin version
- #192: Versioning "copy to"
- #197: Transfer rates
- #199: gin commit: Working without remotes
- #201: Fix file adding in direct mode
- #202: Log trimming
- #203: Progress bars for file operations
- #204: Fix for ls file status in direct mode
- #205: New command: add-remote
- #206: Handle multiple server configurations

## Version 0.16

### Changes
- Logging changes: More useful logging info and command delimitation.
- Relevant help: When a command is given bad arguments, instead of printing the general help/usage info, it now prints the help/usage for that specific command.
- Fix for stuttering/flashing of text during progress printing on Windows.
- Completely redone command line argument handling and better help formatting.
- New command: `gin repoinfo`
  - Prints the information for a single repository on the server.

### Relevant PRs
- #172: Logging changes
- #174: Use git annex add --update for locking
- #176: Print gin-cli version to log on initialisation (of the log)
- #177: Update help text
- #181: Error message and output improvements
- #182: Print output status only when text has changed
- #184: Better command line argument parsing
- #187: Fix for configuration paths with spaces
- #188: New command: repoinfo


## Version 0.15

### Changes
- No longer commits changes when performing an `AnnexPull` (`git annex sync --no-push`).

### Relevant PRs
- #171: No commit on gin download

## Version 0.14

### Changes
- Host SSH key needs to be added to any non-default host configuration and is strictly checked.
- Various improvements and bug fixes.

### Relevant PRs
- #161: Fixed bug where output message for file removals was 'Adding remove'.
- #162: Strict host key checking for git and ssh commands.
- #163: Allow running git and annex proxy commands without token.
- #168: Ignore old (deprecated) config paths.
- #169: Disable symlinks explicitly on Windows.

## Version 0.13

### New features
- Create repository on the server without cloning: `gin create --no-clone`
    - Cannot be used in combination with `--here`.
- Delete public SSH key from the server: `gin keys --delete <index>`

### Relevant PRs
- #157: Delete key command
- #160: `gin create --no-clone`

## Version 0.12

### New features and feature changes
- Create repository out of current directory
    - `gin create --here` will create a new repository on the remote server and initialise the current working directory with the appropriate remotes and annex configuration.
- JSON output: Most commands now support the `--json` flag for JSON formatted output.
- The location for configuration files has changed on all platforms. Platform specific locations are now used. On first run, the client looks for files in the old configuration directories and transfers them to the new location if necessary.
- If a command requires login and the user is not logged in, instead of simply printing an informative message, the user is prompted for login (unless `--json` is specified).
- Local (in-repo) config file can only be used to specify annex selection. In other words, only the `annex.minsize` and `annex.excludes` is used. All other options are read from the global and default configurations.

ginclient/config/config.go
// Package config handles reading of the user configuration for the client.
package config

import (
	"fmt"
	"os"
	"path/filepath"

	"github.com/G-Node/gin-cli/ginclient/log"
	"github.com/fatih/color"
	"github.com/shibukawa/configdir"
	"github.com/spf13/viper"
)

const (
	defaultFileName = "config.yml"
)

var (
	configDirs = configdir.New("g-node", "gin")
	yellow     = color.New(color.FgYellow).SprintFunc()

	ginDefaultServer = ServerCfg{
		WebCfg{
			Protocol: "https",
			Host:     "web.gin.g-node.org",
			Port:     443,
		},
		GitCfg{
			Host:    "gin.g-node.org",
			Port:    22,
			User:    "git",
			HostKey: "gin.g-node.org,141.84.41.216 ecdsa-sha2-nistp256 AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBE5IBgKP3nUryEFaACwY4N3jlqDx8Qw1xAxU2Xpt5V0p9RNefNnedVmnIBV6lA3n+9kT1OSbyqA/+SgsQ57nHo0=",
		},
	}

	defaultConf = map[string]interface{}{
		// Binaries
		"bin.git":      "git",
		"bin.gitannex": "git-annex",
		"bin.ssh":      "ssh",
		// Annex filters
		"annex.minsize": "10M",
		"servers.gin":   ginDefaultServer,
		"defaultserver": "gin",
	}

	// configuration cache: used to avoid rereading during a single command invocation
	configuration GinCliCfg
	set           = false
)

// Types

// WebCfg is the configuration for the web server.
type WebCfg struct {
	Protocol string
	Host     string
	Port     uint16
}

// AddressStr constructs a full address string from the configuration.
// The string has the format Scheme://Host:Port (e.g., https://web.gin.g-node.org:443)
func (c WebCfg) AddressStr() string {
	return fmt.Sprintf("%s://%s:%d", c.Protocol, c.Host, c.Port)
}

// GitCfg is the configuration for the git server.
type GitCfg struct {
	User    string
	Host    string
	Port    uint16
	HostKey string
}

// AddressStr constructs a full address string from the configuration.
// The string has the format ssh://User@Host:Port (e.g., ssh://git@gin.g-node.org:22)
func (c GitCfg) AddressStr() string {
	return fmt.Sprintf("ssh://%s@%s:%d", c.User, c.Host, c.Port)
}

// ServerCfg holds the information required for GIN servers (web and git).
type ServerCfg struct {
	Web WebCfg
	Git GitCfg
}

// GinCliCfg holds the client configuration values.
type GinCliCfg struct {
	Servers       map[string]ServerCfg
	DefaultServer string
	Bin           struct {
		Git      string
		GitAnnex string
		SSH      string
	}
	Annex struct {
		Exclude []string
		MinSize string
	}
}

// Read loads in the configuration from the config file(s), merges any defined values into the default configuration, and returns a populated GinConfiguration struct.
// The configuration is cached. Subsequent reads reuse the already loaded configuration.
func Read() GinCliCfg {
	if set {
		return configuration
	}
	viper.Reset()
	viper.SetTypeByDefaultValue(true)

	for k, v := range defaultConf {
		viper.SetDefault(k, v)
	}

	// Merge in user config file
	confpath, _ := Path(false)
	confpath = filepath.Join(confpath, defaultFileName)
	viper.SetConfigFile(confpath)
	cerr := viper.MergeInConfig()
	if cerr == nil {
		log.Write("Found config file %s", confpath)
	}

	viper.Unmarshal(&configuration)

	removeInvalidServerConfs()

	// configuration file in the repository root (annex excludes and size threshold only)
	reporoot, err := findreporoot(".")
	if err == nil {
		confpath := filepath.Join(reporoot, defaultFileName)
		viper.SetConfigFile(confpath)
		cerr = viper.MergeInConfig()
		if cerr == nil {
			log.Write("Found config file %s", confpath)
		}
	}
	configuration.Annex.Exclude = viper.GetStringSlice("annex.exclude")
	configuration.Annex.MinSize = viper.GetString("annex.minsize")

	set = true
	return configuration
}

func removeInvalidServerConfs() {
	// Check server configurations for invalid names and port numbers
	for alias := range viper.GetStringMap("servers") {
		if alias == "dir" {
			fmt.Fprintf(color.Error, "%s server alias '%s' is not allowed (reserved word): server configuration ignored\n", yellow("[warning]"), alias)
			delete(configuration.Servers, alias)
			continue
		}
		webport := viper.GetInt(fmt.Sprintf("servers.%s.web.port", alias))
		gitport := viper.GetInt(fmt.Sprintf("servers.%s.git.port", alias))
		if webport < 0 || webport > 65535 || gitport < 0 || gitport > 65535 {
			if alias == "gin" {
				fmt.Fprintf(color.Error, "%s invalid value found in configuration for '%s': using default\n", yellow("[warning]"), alias)
				configuration.Servers["gin"] = ginDefaultServer
			} else {
				fmt.Fprintf(color.Error, "%s invalid value found in configuration for '%s': server configuration ignored\n", yellow("[warning]"), alias)
				delete(configuration.Servers, alias)
			}
		}
	}
}

// appendToFile appends a key-value to the configuration file.
// A useful utility function that loads the configuration only from the file, adds the new key-value pair, and saves it back, without loading the built-in defaults.
// On successful write, the read cache is invalidated.
func appendToFile(key string, value interface{}) error {
	// Read in the file configuration ONLY
	confpath, err := Path(true) // create config path if necessary
	if err != nil {
		return err
	}
	confpath = filepath.Join(confpath, defaultFileName)
	v := viper.New()
	v.SetConfigFile(confpath)

	v.ReadInConfig()
	v.Set(key, value)
	v.WriteConfig()
	// invalidate the read cache
	set = false
	return nil
}

// AddServerConf writes a new server configuration into the user config file.
func AddServerConf(alias string, newcfg ServerCfg) error {
	key := fmt.Sprintf("servers.%s", alias)
	return appendToFile(key, newcfg)
}

// RmServerConf removes a server configuration from the user config file.
func RmServerConf(alias string) {
	var c GinCliCfg
	v := viper.New()
	// Merge in user config file
	confpath, _ := Path(false)
	confpath = filepath.Join(confpath, defaultFileName)
	v.SetConfigFile(confpath)
	v.ReadInConfig()
	v.Unmarshal(&c)
	delete(c.Servers, alias)
	v.Set("servers", c.Servers)
	v.WriteConfig()
	set = false
}

// SetDefaultServer writes the given name to the config file to server as the default server for web calls.
// An error is returned if the name doesn't exist in the current configuration.
func SetDefaultServer(alias string) {
	appendToFile("defaultserver", alias)
}

// Path returns the configuration path where configuration files should be stored.
// If the GIN_CONFIG_DIR environment variable is set, its value is returned, otherwise the platform default is used.
// If create is true and the directory does not exist, the full path is created.
func Path(create bool) (string, error) {
	confpath := os.Getenv("GIN_CONFIG_DIR")
	if confpath == "" {
		confpath = configDirs.QueryFolders(configdir.Global)[0].Path
	}
	var err error
	if create {
		err = os.MkdirAll(confpath, 0755)
		if err != nil {
			return "", fmt.Errorf("could not create config directory %s", confpath)
		}
	}
	return confpath, err
}

// Util functions //

// pathExists returns true if the path exists
func pathExists(path string) bool {
	if _, err := os.Stat(path); os.IsNotExist(err) {
		return false
	}
	return true
}

func findreporoot(path string) (string, error) {
	var err error
	path, err = filepath.Abs(path)
	if err != nil {
		return "", err
	}
	gitdir := filepath.Join(path, ".git")
	if pathExists(gitdir) {
		return path, nil
	}
	updir := filepath.Dir(path)
	if updir == path {
		// root reached
		return "", fmt.Errorf("Not a repository")
	}

	return findreporoot(updir)
}
